### ISSUE TITLE
Inlining of mutually recursive functions

### ISSUE THREAD

**Issue by simonster** [JuliaLang/julia#13293]
I'm not sure anyone should particularly care about this but:

```julia
julia> fib(n) = n < 2 ? n : fib2(n-1) + fib2(n-2)
fib (generic function with 1 method)

julia> fib2(n) = n < 2 ? n : fib(n-1) + fib(n-2)
fib2 (generic function with 1 method)

julia> @code_llvm fib(1)

define i64 @julia_fib_44516(i64) {
top:
  %1 = icmp sgt i64 %0, 1
  br i1 %1, label %L, label %if

if:                                               ; preds = %top
  ret i64 %0

L:                                                ; preds = %top
  %2 = add i64 %0, -1
  %3 = call i1 @"julia_<537"(i64 %2, i64 2)
  br i1 %3, label %L3, label %L2

L2:                                               ; preds = %L
  %4 = call i64 @julia_-364(i64 %2, i64 1)
  %5 = call i64 @julia_fib_44516(i64 %4)
  %6 = call i64 @julia_-364(i64 %2, i64 2)
  %7 = call i64 @julia_fib_44516(i64 %6)
  %8 = call i64 @"julia_+315"(i64 %5, i64 %7)
  br label %L3

L3:                                               ; preds = %L2, %L
  %_var1.0 = phi i64 [ %8, %L2 ], [ %2, %L ]
  %9 = add i64 %0, -2
  %10 = call i1 @"julia_<537"(i64 %9, i64 2)
  br i1 %10, label %L7, label %L6

L6:                                               ; preds = %L3
  %11 = call i64 @julia_-364(i64 %9, i64 1)
  %12 = call i64 @julia_fib_44516(i64 %11)
  %13 = call i64 @julia_-364(i64 %9, i64 2)
  %14 = call i64 @julia_fib_44516(i64 %13)
  %15 = call i64 @"julia_+315"(i64 %12, i64 %14)
  br label %L7

L7:                                               ; preds = %L6, %L3
  %_var0.0 = phi i64 [ %15, %L6 ], [ %9, %L3 ]
  %16 = add i64 %_var0.0, %_var1.0
  ret i64 %16
}
```

When the inliner bails out it seems to leave `-`, `+`, and `<` as function calls in addition to `fib`. I noticed this because I was playing around after seeing GCC beats us at `fib` by iteratively inlining the recursive function.

**Comment by simonster** [comment]
Seems to be fixed in 0.5.

**Comment by tkelman** [comment]
will regressions in this be caught by the daily microbenchmarks? assuming people keep a close enough eye on them over time
